text/0016-internationalization.md

```diff
@@ -0,0 +1,117 @@
+    
+2.  The default language is English, this should be made visible and explicit in the code.
+    
+3.  The ESM translation module will be able to translate content **without** connectivity.
+    
+4.  Presently, OpenMRS uses Transifex ([https://www.transifex.com/openmrs/OpenMRS/dashboard/](https://www.transifex.com/openmrs/OpenMRS/dashboard/)) to manage translations. We will continue to use Transifex to support translations for the MF project.
+    
+5.  Possibly find ways to reuse OpenMRS module translations found in the `message.properties` file.
+
+7.  Update mechanism: Once translations are marked as complete, create a mechanism to publish the translated strings to a github repo. 
+    
+
+## Reason for decision
+
+- Most distributions are deployed to areas with no to intermittent connectivity, thus the translations module should be completely offline.
+
+- English is the lingua franca and spoken in most parts of the world.
+
+- Reasons for choosing Transifex:
+	 - Transifex is the most popular internationalization SaaS and offers tools that would make the translations workflow smoother.
+	 - The community is already familiar using transifex for translation services.
+
+
+## Alternatives
+- No translation service at all and default to English for all frontend modules.
+-  Use another service provider built specifically for frontend applications such as [react-i18next](https://github.com/i18next/react-i18next)
+- We could store the translations with each `esm` module instead of having a translations `esm` module.
+- Repo to handle translations service: we could have the translations handled by `openmrs-esm-api`
+
+
+## Common practices (not enforced)
+
+
+### Definitions
+
+Messages - OpenMRS historically used this word to refer to the content that would be replaced.
+
+  
+### Structure of Message File
+
+Here’s an example of how message.properties file looks like:
+
+[https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties](https://github.com/openmrs/openmrs-module-legacyui/blob/master/api/src/main/resources/messages.properties)
+
+
+### Existing API
+
+Here is a [link](https://github.com/openmrs/openmrs-core/blob/master/api/src/main/java/org/openmrs/messagesource/MessageSourceService.java) to the github repo for the interface to the java MessageSourceService in OpenMRS core. It has four methods:
+
+    public String getMessage(String s);
+    
+    public MutableMessageSource getActiveMessageSource();
+    
+    public  void  setActiveMessageSource(MutableMessageSource activeMessageSource);
+    
+    public  void  setMessageSources(Set<MutableMessageSource>  availableMessageSources);
+
+  
+
+### Proposed API
+
+1.  `setPreferredLocale()`: sets the preferred locale of the user, defaults to English if the user has no preferred locale.
+    
+2.  `getMessages(translation)` : takes argument `es-mx` and returns all the messages for that locale.
```

Comment:
Right now the RefApp frontend makes a zillion async API calls for messages, many redundantly. Some of these fail frequently, causing `coreapps.foo.bar.baz` to be all over the page where there should be UI text. I would like to avoid re-creating this problem. Maybe they can be grouped by page or something?

Reply:
Can this be flexible? eg. go to a client side cache/store for these first and then retrieve from server if not found?  Then potentially different consumers could choose to preload the cache or load on demand.

Reply:
Storing in a client side cache could be cool.

I'm thinking that we would make one network request per Transifex *project*. Most definitely not per string.

So there would be an "MF Core" transifex project that would always be available. But then new modules could specify dependencies on their own transifex projects which would be lazily downloaded. Introducing a client-side cache could be interesting, too, to get rid of the need for even those network requests.

Reply:
After thinking about this more, I think that using the browser's built-in cache for network requests would be more than sufficient here. The server for the frontend files can send a `Cache-Control` header and that should be enough to do the trick.

We should be shooting for **a few network requests at most** for translations when a page is loaded. If we get into a scenario where we're making more than 10 network requests just for translations, we've gone wrong.